src/varcache.c

```diff
@@ -88,7 +88,7 @@ static void init_var_lookup_from_config(const char *cf_track_extra_parameters, i
 
 void init_var_lookup(const char *cf_track_extra_parameters)
 {
-	const char *names[] = { "DateStyle", "client_encoding", "TimeZone", "standard_conforming_strings", "application_name", NULL };
+	const char *names[] = { "DateStyle", "client_encoding", "TimeZone", "standard_conforming_strings", "application_name", "in_hot_standby", "default_transaction_read_only", NULL };
```

Comment:
Brian and I were discussing this offline and wondered if we need to make 2 distinct lists. One for values that are readwrite and should be set on the connection startup, and another for readonly values that we want to be able to read from the connection but not attempt to set. Thoughts?

Reply:
That's a good question. I think it would probably make sense. This is also very related to #1205, because that also needs to track these read-only settings to be able to send the new versions to the client if they change on the server side. 

Regarding #1205, currently that PR is sending all the ParameterStatus messages again to every client if a new server reported a different one. I think a more correct, aka more in line with postgres behaviour, would be to track these read-only fields per server, and if a client is linked to a server where the values were different than for the previous link, then we'd send ParameterStatus messages for just those ones. We'd have to be a bit careful with the perf implications of that though, because it adds overhead on every connection.

And I think `default_transaction_read_only` is probably a bit special too in this regard, because it can actually be set by the client. So tracking it and then setting it according to the last client setting could actually make sense. But I'm curious how this interacts with servers configuring this setting and then changing it. If we'd set it based on the client its last value, then only new clients would get the new value. Which probably is fine, because the same would actually apply to normal connection to postgres.

@emelsimsek might have some ideas around this too.

Reply:
After thinking about this more and reading the docs again I'm not sure that it should be a design goal here should be to enforce whether or not a parameter tracked by the varcache is read-only or read-write. If we set the varcache only after getting the pertinent `ParameterStatus` message from the server we can be certain that it is the valid value for the back-end connection.

From the docs (emphasis mine):

```ParameterStatus messages will be generated whenever the active value changes for any of the parameters the backend believes the frontend should know about. Most commonly this occurs in response to a SET SQL command executed by the frontend, and this case is effectively synchronous — but it is also possible for parameter status changes to occur because the administrator changed a configuration file and then sent the SIGHUP signal to the server. Also, *if a SET command is rolled back, an appropriate ParameterStatus message will be generated to report the current effective value.*```

I think that it's none of our business if the client issues a `SET` command that will fail and we can continue to transparently return whatever error the server gave the client.

Reply:
The docs go on to talk about the hard-wired set of `ParameterStatus` messages and warn:

> This set might change in the future, or even become configurable.

So, IMHO, let's leave it.

Reply:
I think that, then, separates this concern from #1205 though I think that propagating `ParameterStatus` out to all connected clients _may_ be a good change. I am not 100% sure, though, because if an administrator issues a `reload` would it not already inform all connected clients?

Reply:
https://www.postgresql.org/docs/current/protocol-flow.html

Reply:
@ya-mouse OK thanks, your comment made me think a bit further.

At a minimum I need to add tests wherein the server role changes and observe what happens to the server connections that no longer satisfy `target_session_attrs` and what the client sees. Will look tonight.